// @tryghost/ghost/build/nestjs/modules/activitypub.module.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ActivityPubModule = void 0;
const common_1 = require("@nestjs/common");
const actor_repository_in_memory_1 = require("../../db/in-memory/actor.repository.in-memory");
const activitypub_controller_1 = require("../../http/frontend/controllers/activitypub.controller");
const webfinger_service_1 = require("../../core/activitypub/webfinger.service");
const jsonld_service_1 = require("../../core/activitypub/jsonld.service");
const webfinger_controller_1 = require("../../http/frontend/controllers/webfinger.controller");
const activity_service_1 = require("../../core/activitypub/activity.service");
const post_repository_knex_1 = require("../../db/knex/post.repository.knex");
let ActivityPubModule = class ActivityPubModule {
};
exports.ActivityPubModule = ActivityPubModule;
exports.ActivityPubModule = ActivityPubModule = __decorate([
    (0, common_1.Module)({
        controllers: [activitypub_controller_1.ActivityPubController, webfinger_controller_1.WebFingerController],
        exports: [],
        providers: [
            {
                provide: 'ActorRepository',
                useClass: actor_repository_in_memory_1.ActorRepositoryInMemory
            },
            {
                provide: 'ActivityService',
                useClass: activity_service_1.ActivityService
            },
            {
                provide: 'PostRepository',
                useClass: post_repository_knex_1.KnexPostRepository
            },
            webfinger_service_1.WebFingerService,
            jsonld_service_1.JSONLDService
        ]
    })
], ActivityPubModule);
