// @tryghost/ghost/build/db/knex/post.repository.knex.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
var __param = (this && this.__param) || function (paramIndex, decorator) {
    return function (target, key) { decorator(target, key, paramIndex); }
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.KnexPostRepository = void 0;
const common_1 = require("@nestjs/common");
let KnexPostRepository = class KnexPostRepository {
    knex;
    urlUtils;
    constructor(knex, urlUtils) {
        this.knex = knex;
        this.urlUtils = urlUtils;
    }
    async getOne(identifier) {
        return this.getOneById(identifier);
    }
    async getOneById(id) {
        const row = await this.knex('posts').where('id', id.toHexString()).first();
        if (!row) {
            return null;
        }
        return {
            id,
            title: row.title,
            html: this.urlUtils.transformReadyToAbsolute(row.html),
            slug: row.slug,
            visibility: row.visibility
        };
    }
};
exports.KnexPostRepository = KnexPostRepository;
exports.KnexPostRepository = KnexPostRepository = __decorate([
    __param(0, (0, common_1.Inject)('knex')),
    __param(1, (0, common_1.Inject)('UrlUtils')),
    __metadata("design:paramtypes", [Object, Object])
], KnexPostRepository);
;
